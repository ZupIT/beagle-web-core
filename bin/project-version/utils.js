const { execSync } = require('child_process')

function versionToArray(version) {
  return version.split('.').map(versionPart => {
    try {
      return Number.parseInt(versionPart)
    } catch {
      return 0
    }
  })
}

function fetchNpmVersion(packageName) {
  return execSync(`npm show ${packageName} version`, { encoding: 'utf8' }).replace('\n', '')
}

function getPackageJson() {
  return require('../../package.json')
}

module.exports = {
  versionToArray,
  fetchNpmVersion,
  getPackageJson,
}
