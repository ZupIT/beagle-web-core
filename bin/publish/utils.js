
/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

const { execSync } = require('child_process')

function versionToArray(version) {
  return version.split('.').map(versionPart => {
    try {
      return Number.parseInt(versionPart)
    } catch {
      return 0
    }
  })
}

function fetchNpmVersion(packageName) {
  return execSync(`npm show ${packageName} version`, { encoding: 'utf8' }).replace('\n', '')
}

function getPackageJson() {
  return require('../../package.json')
}

module.exports = {
  versionToArray,
  fetchNpmVersion,
  getPackageJson,
}
