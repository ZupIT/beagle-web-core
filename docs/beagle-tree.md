# Beagle Tree
The Beagle Tree is the entity used to represent a server-driven view, as its name says, it has a
tree structure. Beagle offers many functions to help traversing this tree, finding specific
components or altering it.
