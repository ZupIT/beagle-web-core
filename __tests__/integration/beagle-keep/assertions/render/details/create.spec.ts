/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import Tree from 'beagle-tree'
import { IdentifiableBeagleUIElement } from 'beagle-tree/types'
import setup from '../../../backend/routes'
import { enableLogging, disableLogging } from '../../../../../utils/log'
import { RenderingResult, renderDetailsView } from '../../utils'

/**
 * This page is very different from details and labels, so we'll test every lifecycle again.
 * Here we test the "create note" mode of the view "details".
 */
describe('Beagle Keep: render details (create note)', () => {
  let renderedTrees: RenderingResult
  enableLogging()
  setup()
  
  beforeAll(async () => {
    renderedTrees = await renderDetailsView()
  })

  afterAll(disableLogging)

  /**
   * Two renders are expected:
   * - first: the view details, with a loading overlay.
   * - second: the same view after the onInit from the root container is run. The onInit will call
   * a setContext action, which will trigger a new render. Since the second render is triggered by
   * a setContext, only a partial render is expected and a not full render, i.e., every lifecycle
   * before the view snapshot is not run.
   */
  it('should do one full render and one partial render with no errors or warnings', () => {
    expect(renderedTrees.beforeStart.length).toBe(1)
    expect(renderedTrees.beforeViewSnapshot.length).toBe(1)
    expect(renderedTrees.afterViewSnapshot.length).toBe(2)
    expect(renderedTrees.beforeRender.length).toBe(2)
    expect(renderedTrees.render.length).toBe(2)
    expect(globalMocks.log).not.toHaveBeenCalled()
  })

  describe('first render of details', () => {
    /**
     * Should start rendering the details page. The snapshot here is raw, just the way the server
     * sent it.
     */
    it('should match snapshot on before start', () => {
      expect(renderedTrees.beforeStart[0]).toMatchSnapshot()
    })

    /**
     * Expected difference from the previous lifecycle (beforeStart): ids for every component.
     * Containers should have special ids according to their beforeStart lifecycle.
     */
    it('should match snapshot on before view snapshot', () => {
      expect(renderedTrees.beforeViewSnapshot[0]).toMatchSnapshot()
    })

    /**
     * Expected difference from the previous lifecycle (beforeViewSnapshot): this page has inputs
     * which are created by the backend with the property "model". "model" is not understood by
     * Beagle, so there's a beforeViewSnapshot lifecycle for text inputs and text areas that
     * transforms the "model" property into a "value" expression and an "onChange" event, which
     * Beagle can understand. Since this transformation occurred in beforeViewSnapshot, the tree
     * received by the lifecycle afterViewSnapshot must have these changes.
     */
    it('should match snapshot on after view snapshot', () => {
      expect(renderedTrees.afterViewSnapshot[0]).toMatchSnapshot()
    })

    /**
     * Expected differences from the previous lifecycle (afterViewSnapshot):
     * - every button should also have the property isSubmit.
     * - actions should have been transformed into functions.
     * - expressions should have been assigned real values.
     * - styles should have been translated to camel-case css.
     */
    it('should match snapshot on before render', () => {
      expect(renderedTrees.beforeRender[0]).toMatchSnapshot()
    })

    /**
     * The tree here is fully processed and ready to be rendered.
     * Expected difference from the previous lifecycle (beforeRender): every container should have
     * { style: { color: '#FFF' } }.
     */
    it('should render details for the first time', () => {
      expect(renderedTrees.render[0]).toMatchSnapshot()
    })
  })

  /**
   * The only difference from the second render to the first is that "isVisible" in the component
   * "custom:loadingOverlay" will be false instead of true.
   */
  describe('second render of details (partial). Hides the loading.', () => {
    function loadingOverlayShouldBeInvisible(details: IdentifiableBeagleUIElement) {
      const loadingOverlay = Tree.findByType(details, 'custom:loadingOverlay')[0]
      expect(loadingOverlay.isVisible).toBe(false)
    }

    function shouldBeTheSameAsPreviousExcludingLoadingOverlayVisibility(
      current: IdentifiableBeagleUIElement,
      previous: IdentifiableBeagleUIElement,
    ) {
      const currentWithoutChanges = Tree.clone(current)
      const loadingOverlay = Tree.findByType(currentWithoutChanges, 'custom:loadingOverlay')[0]
      currentWithoutChanges.context!.value.isLoading = true
      loadingOverlay.isVisible = true
      expect(JSON.stringify(currentWithoutChanges)).toEqual(JSON.stringify(previous))
    }

    /**
     * At this time, the only difference from the previous render is the context, where isLoading is
     * now false.
     */
    it('afterViewSnapshot: isLoading in the context should be false', () => {
      const details = renderedTrees.afterViewSnapshot[1]
      expect(details.context!.value.isLoading).toBe(false)
    })

    it(
      'afterViewSnapshot: with the exception of the context, the rest of the tree should be the same as the last afterViewSnapshot',
      () => {
        const detailsWithoutChanges = Tree.clone(renderedTrees.afterViewSnapshot[1])
        detailsWithoutChanges.context!.value.isLoading = true
        expect(detailsWithoutChanges).toEqual(renderedTrees.afterViewSnapshot[0])
      },
    )

    /**
     * In comparison to the last beforeRender, there are two differences:
     * - isLoading, in the context is false, as seen in the last lifecycle (afterViewSnapshot)
     * - the property "isVisible" from the component "custom:loadingOverlay" is resolved to false.
     */
    it('beforeRender: loading overlay should be invisible', () => {
      const details = renderedTrees.beforeRender[1]
      loadingOverlayShouldBeInvisible(details)
    })

    it(
      'beforeRender: with the exception of the visibility of the loading overlay, the rest of the tree should be the same as the last beforeRender',
      () => {
        const current = renderedTrees.beforeRender[1]
        const previous = renderedTrees.beforeRender[0]
        shouldBeTheSameAsPreviousExcludingLoadingOverlayVisibility(current, previous)
      },
    )

    /**
     * Same differences to the last render as the previous lifecycle (beforeRender)
     */
    it('render: loading overlay should be invisible', () => {
      const details = renderedTrees.render[1]
      loadingOverlayShouldBeInvisible(details)
    })

    it(
      'render: with the exception of the visibility of the loading overlay, the rest of the tree should be the same as the last render',
      () => {
        const current = renderedTrees.render[1]
        const previous = renderedTrees.render[0]
        shouldBeTheSameAsPreviousExcludingLoadingOverlayVisibility(current, previous)
      },
    )
  })
})
