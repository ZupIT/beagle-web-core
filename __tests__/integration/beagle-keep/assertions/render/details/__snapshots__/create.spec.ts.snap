// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Beagle Keep: render details (create note) first render of details should match snapshot on after view snapshot 1`] = `
Object {
  "_beagleComponent_": "beagle:container",
  "children": Array [
    Object {
      "_beagleComponent_": "custom:loadingOverlay",
      "id": "_beagle_1",
      "isVisible": "@{form.isLoading}",
    },
    Object {
      "_beagleComponent_": "beagle:simpleForm",
      "children": Array [
        Object {
          "_beagleComponent_": "beagle:container",
          "children": Array [
            Object {
              "_beagleComponent_": "beagle:textInput",
              "error": "@{condition(isEmpty(form.data.title), 'This field is required', '')}",
              "id": "form:title",
              "onBlur": Array [
                Object {
                  "_beagleAction_": "beagle:setContext",
                  "contextId": "form",
                  "path": "showFieldError.title",
                  "value": true,
                },
              ],
              "onChange": Array [
                Object {
                  "_beagleAction_": "beagle:setContext",
                  "contextId": "form",
                  "path": "data.title",
                  "value": "@{onChange.value}",
                },
              ],
              "placeholder": "Title",
              "showError": "@{or(form.showErrors, form.showFieldError.title)}",
              "value": "@{form.data.title}",
            },
            Object {
              "_beagleComponent_": "custom:textArea",
              "error": "@{condition(isEmpty(form.data.text), 'This field is required', '')}",
              "id": "form:text",
              "label": "Content",
              "onBlur": Array [
                Object {
                  "_beagleAction_": "beagle:setContext",
                  "contextId": "form",
                  "path": "showFieldError.text",
                  "value": true,
                },
              ],
              "onChange": Array [
                Object {
                  "_beagleAction_": "beagle:setContext",
                  "contextId": "form",
                  "path": "data.text",
                  "value": "@{onChange.value}",
                },
              ],
              "showError": "@{or(form.showErrors, form.showFieldError.text)}",
              "value": "@{form.data.text}",
            },
          ],
          "id": "container:1",
          "style": Object {
            "backgroundColor": "#333",
            "cornerRadius": Object {
              "radius": 20,
            },
            "flex": Object {
              "flexDirection": "COLUMN",
            },
            "margin": Object {
              "bottom": Object {
                "type": "REAL",
                "value": 30,
              },
            },
            "padding": Object {
              "all": Object {
                "type": "REAL",
                "value": 15,
              },
            },
            "size": Object {
              "maxWidth": Object {
                "type": "REAL",
                "value": 500,
              },
            },
          },
        },
        Object {
          "_beagleComponent_": "beagle:container",
          "children": Array [
            Object {
              "_beagleComponent_": "beagle:button",
              "id": "_beagle_3",
              "onPress": Array [
                Object {
                  "_beagleAction_": "beagle:popView",
                },
              ],
              "text": "Go back",
            },
            Object {
              "_beagleComponent_": "beagle:button",
              "id": "_beagle_4",
              "onPress": Array [
                Object {
                  "_beagleAction_": "beagle:submitForm",
                },
              ],
              "text": "Save",
            },
          ],
          "id": "container:2",
          "style": Object {
            "flex": Object {
              "flex": 1,
              "flexDirection": "ROW",
              "justifyContent": "SPACE_BETWEEN",
            },
          },
        },
      ],
      "id": "_beagle_2",
      "onSubmit": Array [
        Object {
          "_beagleAction_": "beagle:setContext",
          "contextId": "form",
          "path": "isLoading",
          "value": true,
        },
        Object {
          "_beagleAction_": "beagle:sendRequest",
          "data": "@{form.data}",
          "method": "@{condition(isNull(global.selectedNote), 'POST', 'PUT')}",
          "onError": Array [
            Object {
              "_beagleAction_": "custom:feedback",
              "message": "Could not save the message. Please, try again later.",
              "type": "error",
            },
          ],
          "onFinish": Array [
            Object {
              "_beagleAction_": "beagle:setContext",
              "contextId": "form",
              "path": "isLoading",
              "value": false,
            },
          ],
          "onSuccess": Array [
            Object {
              "_beagleAction_": "beagle:condition",
              "condition": "@{sNull(global.selectedNote)}",
              "onFalse": Array [
                Object {
                  "_beagleAction_": "custom:feedback",
                  "message": "Note successfully updated!",
                  "type": "success",
                },
              ],
              "onTrue": Array [
                Object {
                  "_beagleAction_": "custom:feedback",
                  "message": "Note successfully created!",
                  "type": "success",
                },
                Object {
                  "_beagleAction_": "beagle:popView",
                },
              ],
            },
          ],
          "url": "https://keep.beagle.io/note",
        },
      ],
    },
  ],
  "context": Object {
    "id": "form",
    "value": Object {
      "data": Object {
        "labels": Array [],
        "text": "",
        "title": "",
      },
      "isLoading": true,
      "labels": Array [
        Object {
          "color": "#00ff00",
          "id": 0,
          "name": "Personal",
        },
        Object {
          "color": "#0000ff",
          "id": 1,
          "name": "Work",
        },
        Object {
          "color": "#ff0000",
          "id": 2,
          "name": "Urgent",
        },
        Object {
          "color": "#ffff00",
          "id": 3,
          "name": "Supermarket",
        },
      ],
    },
  },
  "id": "container:0",
  "onInit": Array [
    Object {
      "_beagleAction_": "beagle:condition",
      "condition": "@{isNull(global.selectedNote)}",
      "onFalse": Array [
        Object {
          "_beagleAction_": "beagle:sendRequest",
          "onError": Array [
            Object {
              "_beagleAction_": "beagle:addChildren",
              "mode": "replace",
              "value": Object {
                "_beagleComponent_": "custom:error",
                "message": "Couldn't find note with id @{global.selectedNote}",
              },
            },
          ],
          "onFinish": Array [
            Object {
              "_beagleAction_": "beagle:setContext",
              "contextId": "form",
              "path": "isLoading",
              "value": false,
            },
          ],
          "onSuccess": Array [
            Object {
              "_beagleAction_": "beagle:setContext",
              "contextId": "form",
              "path": "data",
              "value": Object {
                "id": "@{onSuccess.data.id}",
                "labels": "@{onSuccess.data.labels}",
                "text": "@{onSuccess.data.text}",
                "title": "@{onSuccess.data.title}",
              },
            },
          ],
          "url": "https://keep.beagle.io/note/@{global.selectedNote}",
        },
      ],
      "onTrue": Array [
        Object {
          "_beagleAction_": "beagle:setContext",
          "contextId": "form",
          "path": "isLoading",
          "value": false,
        },
      ],
    },
  ],
  "style": Object {
    "flex": Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    },
    "size": Object {
      "height": Object {
        "type": "PERCENT",
        "value": 100,
      },
    },
  },
  "version": "Beagle Keep v1.0",
}
`;

exports[`Beagle Keep: render details (create note) first render of details should match snapshot on before render 1`] = `
Object {
  "_beagleComponent_": "beagle:container",
  "children": Array [
    Object {
      "_beagleComponent_": "custom:loadingOverlay",
      "id": "_beagle_1",
      "isVisible": true,
    },
    Object {
      "_beagleComponent_": "beagle:simpleForm",
      "children": Array [
        Object {
          "_beagleComponent_": "beagle:container",
          "children": Array [
            Object {
              "_beagleComponent_": "beagle:textInput",
              "error": "This field is required",
              "id": "form:title",
              "onBlur": [Function],
              "onChange": [Function],
              "placeholder": "Title",
              "showError": false,
              "value": "",
            },
            Object {
              "_beagleComponent_": "custom:textArea",
              "error": "This field is required",
              "id": "form:text",
              "label": "Content",
              "onBlur": [Function],
              "onChange": [Function],
              "showError": false,
              "value": "",
            },
          ],
          "id": "container:1",
          "style": Object {
            "backgroundColor": "#333",
            "borderRadius": "40px",
            "flexDirection": "column",
            "marginBottom": "30px",
            "maxWidth": "500px",
            "padding": "15px",
          },
        },
        Object {
          "_beagleComponent_": "beagle:container",
          "children": Array [
            Object {
              "_beagleComponent_": "beagle:button",
              "id": "_beagle_3",
              "isSubmit": false,
              "onPress": [Function],
              "text": "Go back",
            },
            Object {
              "_beagleComponent_": "beagle:button",
              "id": "_beagle_4",
              "isSubmit": true,
              "onPress": [Function],
              "text": "Save",
            },
          ],
          "id": "container:2",
          "style": Object {
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
          },
        },
      ],
      "id": "_beagle_2",
      "onSubmit": [Function],
    },
  ],
  "context": Object {
    "id": "form",
    "value": Object {
      "data": Object {
        "labels": Array [],
        "text": "",
        "title": "",
      },
      "isLoading": true,
      "labels": Array [
        Object {
          "color": "#00ff00",
          "id": 0,
          "name": "Personal",
        },
        Object {
          "color": "#0000ff",
          "id": 1,
          "name": "Work",
        },
        Object {
          "color": "#ff0000",
          "id": 2,
          "name": "Urgent",
        },
        Object {
          "color": "#ffff00",
          "id": 3,
          "name": "Supermarket",
        },
      ],
    },
  },
  "id": "container:0",
  "onInit": [Function],
  "style": Object {
    "alignItems": "center",
    "flex": 1,
    "height": "100%",
    "justifyContent": "center",
  },
  "version": "Beagle Keep v1.0",
}
`;

exports[`Beagle Keep: render details (create note) first render of details should match snapshot on before start 1`] = `
Object {
  "_beagleComponent_": "beagle:container",
  "children": Array [
    Object {
      "_beagleComponent_": "custom:loadingOverlay",
      "isVisible": "@{form.isLoading}",
    },
    Object {
      "_beagleComponent_": "beagle:simpleForm",
      "children": Array [
        Object {
          "_beagleComponent_": "beagle:container",
          "children": Array [
            Object {
              "_beagleComponent_": "beagle:textInput",
              "error": "@{condition(isEmpty(form.data.title), 'This field is required', '')}",
              "id": "form:title",
              "model": "form.data.title",
              "onBlur": Array [
                Object {
                  "_beagleAction_": "beagle:setContext",
                  "contextId": "form",
                  "path": "showFieldError.title",
                  "value": true,
                },
              ],
              "placeholder": "Title",
              "showError": "@{or(form.showErrors, form.showFieldError.title)}",
            },
            Object {
              "_beagleComponent_": "custom:textArea",
              "error": "@{condition(isEmpty(form.data.text), 'This field is required', '')}",
              "id": "form:text",
              "label": "Content",
              "model": "form.data.text",
              "onBlur": Array [
                Object {
                  "_beagleAction_": "beagle:setContext",
                  "contextId": "form",
                  "path": "showFieldError.text",
                  "value": true,
                },
              ],
              "showError": "@{or(form.showErrors, form.showFieldError.text)}",
            },
          ],
          "style": Object {
            "backgroundColor": "#333",
            "cornerRadius": Object {
              "radius": 20,
            },
            "flex": Object {
              "flexDirection": "COLUMN",
            },
            "margin": Object {
              "bottom": Object {
                "type": "REAL",
                "value": 30,
              },
            },
            "padding": Object {
              "all": Object {
                "type": "REAL",
                "value": 15,
              },
            },
            "size": Object {
              "maxWidth": Object {
                "type": "REAL",
                "value": 500,
              },
            },
          },
        },
        Object {
          "_beagleComponent_": "beagle:container",
          "children": Array [
            Object {
              "_beagleComponent_": "beagle:button",
              "onPress": Array [
                Object {
                  "_beagleAction_": "beagle:popView",
                },
              ],
              "text": "Go back",
            },
            Object {
              "_beagleComponent_": "beagle:button",
              "onPress": Array [
                Object {
                  "_beagleAction_": "beagle:submitForm",
                },
              ],
              "text": "Save",
            },
          ],
          "style": Object {
            "flex": Object {
              "flex": 1,
              "flexDirection": "ROW",
              "justifyContent": "SPACE_BETWEEN",
            },
          },
        },
      ],
      "onSubmit": Array [
        Object {
          "_beagleAction_": "beagle:setContext",
          "contextId": "form",
          "path": "isLoading",
          "value": true,
        },
        Object {
          "_beagleAction_": "beagle:sendRequest",
          "data": "@{form.data}",
          "method": "@{condition(isNull(global.selectedNote), 'POST', 'PUT')}",
          "onError": Array [
            Object {
              "_beagleAction_": "custom:feedback",
              "message": "Could not save the message. Please, try again later.",
              "type": "error",
            },
          ],
          "onFinish": Array [
            Object {
              "_beagleAction_": "beagle:setContext",
              "contextId": "form",
              "path": "isLoading",
              "value": false,
            },
          ],
          "onSuccess": Array [
            Object {
              "_beagleAction_": "beagle:condition",
              "condition": "@{sNull(global.selectedNote)}",
              "onFalse": Array [
                Object {
                  "_beagleAction_": "custom:feedback",
                  "message": "Note successfully updated!",
                  "type": "success",
                },
              ],
              "onTrue": Array [
                Object {
                  "_beagleAction_": "custom:feedback",
                  "message": "Note successfully created!",
                  "type": "success",
                },
                Object {
                  "_beagleAction_": "beagle:popView",
                },
              ],
            },
          ],
          "url": "https://keep.beagle.io/note",
        },
      ],
    },
  ],
  "context": Object {
    "id": "form",
    "value": Object {
      "data": Object {
        "labels": Array [],
        "text": "",
        "title": "",
      },
      "isLoading": true,
      "labels": Array [
        Object {
          "color": "#00ff00",
          "id": 0,
          "name": "Personal",
        },
        Object {
          "color": "#0000ff",
          "id": 1,
          "name": "Work",
        },
        Object {
          "color": "#ff0000",
          "id": 2,
          "name": "Urgent",
        },
        Object {
          "color": "#ffff00",
          "id": 3,
          "name": "Supermarket",
        },
      ],
    },
  },
  "onInit": Array [
    Object {
      "_beagleAction_": "beagle:condition",
      "condition": "@{isNull(global.selectedNote)}",
      "onFalse": Array [
        Object {
          "_beagleAction_": "beagle:sendRequest",
          "onError": Array [
            Object {
              "_beagleAction_": "beagle:addChildren",
              "mode": "replace",
              "value": Object {
                "_beagleComponent_": "custom:error",
                "message": "Couldn't find note with id @{global.selectedNote}",
              },
            },
          ],
          "onFinish": Array [
            Object {
              "_beagleAction_": "beagle:setContext",
              "contextId": "form",
              "path": "isLoading",
              "value": false,
            },
          ],
          "onSuccess": Array [
            Object {
              "_beagleAction_": "beagle:setContext",
              "contextId": "form",
              "path": "data",
              "value": Object {
                "id": "@{onSuccess.data.id}",
                "labels": "@{onSuccess.data.labels}",
                "text": "@{onSuccess.data.text}",
                "title": "@{onSuccess.data.title}",
              },
            },
          ],
          "url": "https://keep.beagle.io/note/@{global.selectedNote}",
        },
      ],
      "onTrue": Array [
        Object {
          "_beagleAction_": "beagle:setContext",
          "contextId": "form",
          "path": "isLoading",
          "value": false,
        },
      ],
    },
  ],
  "style": Object {
    "flex": Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    },
    "size": Object {
      "height": Object {
        "type": "PERCENT",
        "value": 100,
      },
    },
  },
  "version": "",
}
`;

exports[`Beagle Keep: render details (create note) first render of details should match snapshot on before view snapshot 1`] = `
Object {
  "_beagleComponent_": "beagle:container",
  "children": Array [
    Object {
      "_beagleComponent_": "custom:loadingOverlay",
      "id": "_beagle_1",
      "isVisible": "@{form.isLoading}",
    },
    Object {
      "_beagleComponent_": "beagle:simpleForm",
      "children": Array [
        Object {
          "_beagleComponent_": "beagle:container",
          "children": Array [
            Object {
              "_beagleComponent_": "beagle:textInput",
              "error": "@{condition(isEmpty(form.data.title), 'This field is required', '')}",
              "id": "form:title",
              "model": "form.data.title",
              "onBlur": Array [
                Object {
                  "_beagleAction_": "beagle:setContext",
                  "contextId": "form",
                  "path": "showFieldError.title",
                  "value": true,
                },
              ],
              "placeholder": "Title",
              "showError": "@{or(form.showErrors, form.showFieldError.title)}",
            },
            Object {
              "_beagleComponent_": "custom:textArea",
              "error": "@{condition(isEmpty(form.data.text), 'This field is required', '')}",
              "id": "form:text",
              "label": "Content",
              "model": "form.data.text",
              "onBlur": Array [
                Object {
                  "_beagleAction_": "beagle:setContext",
                  "contextId": "form",
                  "path": "showFieldError.text",
                  "value": true,
                },
              ],
              "showError": "@{or(form.showErrors, form.showFieldError.text)}",
            },
          ],
          "id": "container:1",
          "style": Object {
            "backgroundColor": "#333",
            "cornerRadius": Object {
              "radius": 20,
            },
            "flex": Object {
              "flexDirection": "COLUMN",
            },
            "margin": Object {
              "bottom": Object {
                "type": "REAL",
                "value": 30,
              },
            },
            "padding": Object {
              "all": Object {
                "type": "REAL",
                "value": 15,
              },
            },
            "size": Object {
              "maxWidth": Object {
                "type": "REAL",
                "value": 500,
              },
            },
          },
        },
        Object {
          "_beagleComponent_": "beagle:container",
          "children": Array [
            Object {
              "_beagleComponent_": "beagle:button",
              "id": "_beagle_3",
              "onPress": Array [
                Object {
                  "_beagleAction_": "beagle:popView",
                },
              ],
              "text": "Go back",
            },
            Object {
              "_beagleComponent_": "beagle:button",
              "id": "_beagle_4",
              "onPress": Array [
                Object {
                  "_beagleAction_": "beagle:submitForm",
                },
              ],
              "text": "Save",
            },
          ],
          "id": "container:2",
          "style": Object {
            "flex": Object {
              "flex": 1,
              "flexDirection": "ROW",
              "justifyContent": "SPACE_BETWEEN",
            },
          },
        },
      ],
      "id": "_beagle_2",
      "onSubmit": Array [
        Object {
          "_beagleAction_": "beagle:setContext",
          "contextId": "form",
          "path": "isLoading",
          "value": true,
        },
        Object {
          "_beagleAction_": "beagle:sendRequest",
          "data": "@{form.data}",
          "method": "@{condition(isNull(global.selectedNote), 'POST', 'PUT')}",
          "onError": Array [
            Object {
              "_beagleAction_": "custom:feedback",
              "message": "Could not save the message. Please, try again later.",
              "type": "error",
            },
          ],
          "onFinish": Array [
            Object {
              "_beagleAction_": "beagle:setContext",
              "contextId": "form",
              "path": "isLoading",
              "value": false,
            },
          ],
          "onSuccess": Array [
            Object {
              "_beagleAction_": "beagle:condition",
              "condition": "@{sNull(global.selectedNote)}",
              "onFalse": Array [
                Object {
                  "_beagleAction_": "custom:feedback",
                  "message": "Note successfully updated!",
                  "type": "success",
                },
              ],
              "onTrue": Array [
                Object {
                  "_beagleAction_": "custom:feedback",
                  "message": "Note successfully created!",
                  "type": "success",
                },
                Object {
                  "_beagleAction_": "beagle:popView",
                },
              ],
            },
          ],
          "url": "https://keep.beagle.io/note",
        },
      ],
    },
  ],
  "context": Object {
    "id": "form",
    "value": Object {
      "data": Object {
        "labels": Array [],
        "text": "",
        "title": "",
      },
      "isLoading": true,
      "labels": Array [
        Object {
          "color": "#00ff00",
          "id": 0,
          "name": "Personal",
        },
        Object {
          "color": "#0000ff",
          "id": 1,
          "name": "Work",
        },
        Object {
          "color": "#ff0000",
          "id": 2,
          "name": "Urgent",
        },
        Object {
          "color": "#ffff00",
          "id": 3,
          "name": "Supermarket",
        },
      ],
    },
  },
  "id": "container:0",
  "onInit": Array [
    Object {
      "_beagleAction_": "beagle:condition",
      "condition": "@{isNull(global.selectedNote)}",
      "onFalse": Array [
        Object {
          "_beagleAction_": "beagle:sendRequest",
          "onError": Array [
            Object {
              "_beagleAction_": "beagle:addChildren",
              "mode": "replace",
              "value": Object {
                "_beagleComponent_": "custom:error",
                "message": "Couldn't find note with id @{global.selectedNote}",
              },
            },
          ],
          "onFinish": Array [
            Object {
              "_beagleAction_": "beagle:setContext",
              "contextId": "form",
              "path": "isLoading",
              "value": false,
            },
          ],
          "onSuccess": Array [
            Object {
              "_beagleAction_": "beagle:setContext",
              "contextId": "form",
              "path": "data",
              "value": Object {
                "id": "@{onSuccess.data.id}",
                "labels": "@{onSuccess.data.labels}",
                "text": "@{onSuccess.data.text}",
                "title": "@{onSuccess.data.title}",
              },
            },
          ],
          "url": "https://keep.beagle.io/note/@{global.selectedNote}",
        },
      ],
      "onTrue": Array [
        Object {
          "_beagleAction_": "beagle:setContext",
          "contextId": "form",
          "path": "isLoading",
          "value": false,
        },
      ],
    },
  ],
  "style": Object {
    "flex": Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    },
    "size": Object {
      "height": Object {
        "type": "PERCENT",
        "value": 100,
      },
    },
  },
  "version": "Beagle Keep v1.0",
}
`;

exports[`Beagle Keep: render details (create note) first render of details should render details for the first time 1`] = `
Object {
  "_beagleComponent_": "beagle:container",
  "children": Array [
    Object {
      "_beagleComponent_": "custom:loadingOverlay",
      "id": "_beagle_1",
      "isVisible": true,
    },
    Object {
      "_beagleComponent_": "beagle:simpleForm",
      "children": Array [
        Object {
          "_beagleComponent_": "beagle:container",
          "children": Array [
            Object {
              "_beagleComponent_": "beagle:textInput",
              "error": "This field is required",
              "id": "form:title",
              "onBlur": [Function],
              "onChange": [Function],
              "placeholder": "Title",
              "showError": false,
              "value": "",
            },
            Object {
              "_beagleComponent_": "custom:textArea",
              "error": "This field is required",
              "id": "form:text",
              "label": "Content",
              "onBlur": [Function],
              "onChange": [Function],
              "showError": false,
              "value": "",
            },
          ],
          "id": "container:1",
          "style": Object {
            "backgroundColor": "#333",
            "borderRadius": "40px",
            "color": "#FFF",
            "flexDirection": "column",
            "marginBottom": "30px",
            "maxWidth": "500px",
            "padding": "15px",
          },
        },
        Object {
          "_beagleComponent_": "beagle:container",
          "children": Array [
            Object {
              "_beagleComponent_": "beagle:button",
              "id": "_beagle_3",
              "isSubmit": false,
              "onPress": [Function],
              "text": "Go back",
            },
            Object {
              "_beagleComponent_": "beagle:button",
              "id": "_beagle_4",
              "isSubmit": true,
              "onPress": [Function],
              "text": "Save",
            },
          ],
          "id": "container:2",
          "style": Object {
            "color": "#FFF",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "space-between",
          },
        },
      ],
      "id": "_beagle_2",
      "onSubmit": [Function],
    },
  ],
  "context": Object {
    "id": "form",
    "value": Object {
      "data": Object {
        "labels": Array [],
        "text": "",
        "title": "",
      },
      "isLoading": true,
      "labels": Array [
        Object {
          "color": "#00ff00",
          "id": 0,
          "name": "Personal",
        },
        Object {
          "color": "#0000ff",
          "id": 1,
          "name": "Work",
        },
        Object {
          "color": "#ff0000",
          "id": 2,
          "name": "Urgent",
        },
        Object {
          "color": "#ffff00",
          "id": 3,
          "name": "Supermarket",
        },
      ],
    },
  },
  "id": "container:0",
  "onInit": [Function],
  "style": Object {
    "alignItems": "center",
    "color": "#FFF",
    "flex": 1,
    "height": "100%",
    "justifyContent": "center",
  },
  "version": "Beagle Keep v1.0",
}
`;
